packages/discovery/src/discovery/analysis/TemplateService.ts

```diff
@@ -52,33 +65,61 @@ export class TemplateService {
           })
             .filter((x) => x.isFile() && x.name.endsWith('.sol'))
             .map((x) => join(shapePath, x.name))
+      const criteriaPath = join(shapePath, 'criteria.json')
+      const criteria = existsSync(criteriaPath)
+        ? JSON.parse(readFileSync(criteriaPath, 'utf8'))
+        : undefined
+      criteria?.validAddresses?.map((a: string) => EthereumAddress(a))
 
       const templateId = path.substring(resolvedRootPath.length + 1)
-      result[templateId] = solidityShapeFiles
+      result[templateId] = { criteria, paths: solidityShapeFiles }
     }
     return result
   }
 
-  findMatchingTemplates(sources: ContractSources): string[] {
-    const result: string[] = []
-    if (!sources.isVerified) {
-      return result
-    }
-
-    const needleSource = flattenFirstSource(sources)
-    if (needleSource === undefined) {
-      return result
+  findMatchingTemplates(
+    sources: ContractSources,
+    address: EthereumAddress,
+  ): string[] {
+    const sourceHash = hashFirstSource(sources)
+    if (sourceHash === undefined) {
+      return []
     }
+    return this.findMatchingTemplatesByHash(sourceHash, address)
+  }
 
-    const needleHash = sha2_256bit(formatIntoHashable(needleSource))
-    const allShapes = this.getAllShapeHashes()
-    for (const [templateId, haystackHashes] of Object.entries(allShapes)) {
-      if (haystackHashes.includes(needleHash)) {
-        result.push(templateId)
+  findMatchingTemplatesByHash(
+    sourcesHash: Hash256,
+    address: EthereumAddress,
+  ): string[] {
+    const result: [string, number][] = []
+
+    const allShapes = this.getAllShapes()
+    for (const [templateId, shape] of Object.entries(allShapes)) {
+      const criteriaMathes: string[] = []
+      if (shape.criteria && shape.criteria.validAddresses) {
+        if (!shape.criteria.validAddresses.includes(address)) {
+          continue
+        } else {
+          criteriaMathes.push('validAddress')
+        }
+      }
+      if (shape.hashes.includes(sourcesHash)) {
+        criteriaMathes.push('implementation')
+        result.push([templateId, criteriaMathes.length])
       }
     }
 
-    return result
+    const maxMatches = Math.max(...result.map(([, matches]) => matches))
```

Comment:
I think it needs to be done this way because we need to show an error if more than one template matches (with the same number of criteria met). So we can't just pick from the top.